stonesoup/dataassociator/neighbour.py

```diff
@@ -106,3 +110,108 @@ def associate(self, tracks, detections, time):
         associations = max(joint_hypotheses)
 
         return associations
+
+
+class GNNWith2DAssignment(DataAssociator):
+    """Global Nearest Neighbour Associator
+
+    Associates detections to a predicted state using the
+    Global Nearest Neighbour method, utilising a 2D matrix of
+    distances and a "shortest path" assignment algorithm.
+    """
+
+    hypothesiser = Property(
+        Hypothesiser,
+        doc="Generate a set of hypotheses for each prediction-detection pair")
+
+    def associate(self, tracks, detections, time):
+        """Associate a set of detections with predicted states.
+
+        Parameters
+        ----------
+        tracks : set of :class:`Track`
+            Current tracked objects
+        detections : set of :class:`Detection`
+            Retrieved measurements
+        time : datetime
+            Detection time to predict to
+
+        Returns
+        -------
+        dict
+            Key value pair of tracks with associated detection
+        """
+
+        associations = {}
+
+        # Convert sets to indexable lists
+        tracks = list(tracks)
+        detections = list(detections)
+
+        # Generate 2d array "matrix" of hypotheses mapping track to detection
+        hypothesis_matrix = np.empty(
+            [len(tracks), len(detections) + len(tracks)],
+            SingleHypothesis)
+        for i in range(len(tracks)):
+            row = np.empty(
+                [len(detections) + len(tracks)], SingleHypothesis)
+            track_hypotheses = self.hypothesiser.hypothesise(
```

Comment:
It might be worth using the standard way of generating the hypotheses prior to the for loop, i.e.
```python
# Generate a set of hypotheses for each track on each detection
hypotheses = {
    track: self.hypothesiser.hypothesise(track, detections, time)
    for track in tracks}
```

and then this line can be replaced by:
```python
track_hypotheses = hypotheses[track]
```

This will make the hypothesis generation clearer (as it follows the same format as other data associators), while also provisioning for easier integration of changes introduced by #145.